ecs-agent/acs/session/payload_responder.go

```diff
@@ -0,0 +1,162 @@
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package session
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+)
+
+const (
+	PayloadMessageName = "PayloadMessage"
+)
+
+type TaskMessageHandler interface {
+	ProcessTaskMessage(message *ecsacs.PayloadMessage,
+		ackFunc func(*ecsacs.AckRequest, []*ecsacs.IAMRoleCredentialsAckRequest)) error
+}
+
+// payloadResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.PayloadMessage messages sent by ACS.
+type payloadResponder struct {
+	taskMessageHandler TaskMessageHandler
+	respond            wsclient.RespondFunc
+}
+
+// NewPayloadResponder returns an instance of the payloadResponder struct.
+func NewPayloadResponder(taskMessageHandler TaskMessageHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &payloadResponder{
+		taskMessageHandler: taskMessageHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*payloadResponder) Name() string { return "payload responder" }
+
+func (r *payloadResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handlePayloadMessage
+}
+
+func (r *payloadResponder) handlePayloadMessage(message *ecsacs.PayloadMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", PayloadMessageName))
+
+	// Validate fields in the message.
+	if err := validatePayloadMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", PayloadMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle payload message.
+	err := r.taskMessageHandler.ProcessTaskMessage(message, r.ackFunc)
+	if err != nil {
+		logger.Critical(fmt.Sprintf("Unable to handle %s", PayloadMessageName), logger.Fields{
+			field.MessageID: aws.StringValue(message.MessageId),
+			field.Error:     err,
+		})
+	}
+}
+
+// ackFunc sends ACKs of the payload message and of the credentials associated with the tasks contained in the payload
+// message.
+func (r *payloadResponder) ackFunc(payloadAck *ecsacs.AckRequest, credsAcks []*ecsacs.IAMRoleCredentialsAckRequest) {
+	go r.sendAck(payloadAck)
+
+	for _, credsAck := range credsAcks {
+		go r.sendAck(credsAck)
+	}
+}
+
+// sendAck handles the sending of an individual specific ACK, assuming it is of type
+// ecsacs.IAMRoleCredentialsAckRequest or ecsacs.AckRequest.
+func (r *payloadResponder) sendAck(ackRequest interface{}) {
```

Comment:
I see (by looking up the code) that `ecsacs.IAMRoleCredentialsAckRequest` and `ecsacs.AckRequest` are completely different types (only `messageID` in common), but it can be intuitive to think the former one is a "child type" of the later one. Can you add a note here that they are two different things, so the two different type cast we did here makes more sense? (it explains why we need a `interface{}` here)

Reply:
I'll add a note here explaining that `ecsacs.IAMRoleCredentialsAckRequest` and `ecsacs.AckRequest` are two different things to provide more context about the usage of type assertion. Thanks!

Reply:
thanks!